vendor/k8s.io/kubernetes/pkg/client/restclient/request.go

```diff
@@ -793,70 +793,114 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
+	requestChan := make(chan error)
 
-		done := func() bool {
-			// Ensure the response body is fully read and closed
-			// before we reconnect, so that we reuse the same TCP
-			// connection.
-			defer func() {
-				const maxBodySlurpSize = 2 << 10
-				if resp.ContentLength <= maxBodySlurpSize {
-					io.Copy(ioutil.Discard, &io.LimitedReader{R: resp.Body, N: maxBodySlurpSize})
-				}
-				resp.Body.Close()
-			}()
+	// closing this channel terminates the current request once a timeout is exceeded
+	requestCancelChan := make(chan struct{})
+
+	go func() {
+		defer close(requestChan)
+
+		// Right now we make about ten retry attempts if we get a Retry-After response.
+		maxRetries := 10
+		retries := 0
+		for {
+			url := r.URL().String()
+			req, err := http.NewRequest(r.verb, url, r.body)
+			if err != nil {
+				requestChan <- err
+				return
+			}
+			req.Header = r.headers
+			req.Cancel = requestCancelChan
+
+			r.backoffMgr.Sleep(r.backoffMgr.CalculateBackoff(r.URL()))
+			resp, err := client.Do(req)
+			updateURLMetrics(r, resp, err)
+			if err != nil {
+				r.backoffMgr.UpdateBackoff(r.URL(), err, 0)
+			} else {
+				r.backoffMgr.UpdateBackoff(r.URL(), err, resp.StatusCode)
+			}
+			if err != nil {
+				requestChan <- err
+				return
+			}
 
-			retries++
-			if seconds, wait := checkWait(resp); wait && retries < maxRetries {
-				if seeker, ok := r.body.(io.Seeker); ok && r.body != nil {
-					_, err := seeker.Seek(0, 0)
-					if err != nil {
-						glog.V(4).Infof("Could not retry request, can't Seek() back to beginning of body for %T", r.body)
-						fn(req, resp)
-						return true
+			done := func() bool {
+				// Ensure the response body is fully read and closed
+				// before we reconnect, so that we reuse the same TCP
+				// connection.
+				defer func() {
+					const maxBodySlurpSize = 2 << 10
+					if resp.ContentLength <= maxBodySlurpSize {
+						io.Copy(ioutil.Discard, &io.LimitedReader{R: resp.Body, N: maxBodySlurpSize})
+					}
+					resp.Body.Close()
+				}()
+
+				retries++
+				if seconds, wait := checkWait(resp); wait && retries < maxRetries {
+					if seeker, ok := r.body.(io.Seeker); ok && r.body != nil {
+						_, err := seeker.Seek(0, 0)
+						if err != nil {
+							glog.V(4).Infof("Could not retry request, can't Seek() back to beginning of body for %T", r.body)
+							fn(req, resp)
+							return true
+						}
 					}
+
+					glog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", seconds, retries, url)
+					r.backoffMgr.Sleep(time.Duration(seconds) * time.Second)
+					return false
 				}
+				fn(req, resp)
+				return true
+			}()
+			if done {
+				requestChan <- nil
+				return
+			}
+		}
+	}()
 
-				glog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", seconds, retries, url)
-				r.backoffMgr.Sleep(time.Duration(seconds) * time.Second)
-				return false
+	// ignore timeout if timeout value is zero
+	if r.timeout == 0 {
```

Comment:
PTAL at this implementation, it works as expected, however maybe there is a cleaner way to "time out" requests?